### ISSUE TITLE
[Perf][B300] Evaluate compact deterministic RWKV7 parameter-gradient reductions

### ISSUE THREAD

**Issue by heiheiha798** [heiheiha798/flash-linear-attention#4]
## Summary

This issue tracks an unmeasured performance hypothesis for the RWKV7 training path on NVIDIA B300: replace two full-sequence fp32 persistent-parameter gradient buffers with compact deterministic hierarchical reductions.

The frozen source baseline is `e47d5d20aeb5989b58a3738b872e7c288a9fb75f` from the latest `fla-org/flash-linear-attention` `main`. Any speedup, traffic reduction, or model impact below is an acceptance target to measure, not an observed result.

## Bottleneck hypothesis

Every training layer calls both auxiliary ops from `RWKV7Attention.forward`:

- `fla/ops/rwkv7/fused_k_update.py::k_update_bwd` allocates an fp32 `dka_tmp` with the full `k` shape, writes one contribution per token/channel, and then performs `sum(dim=(0, 1))` only to produce the persistent `k_a` gradient.
- `fla/ops/rwkv7/gate_output_correction.py::gate_output_correction_backward_triton` allocates an fp32 `grad_r_k` with the full `r` shape, writes one contribution per token/head/channel, and then performs the analogous sum for persistent `r_k`.

At the public default hidden size 2048 (`H=32`, `D=64`) with `B=4`, `T=4096`, each temporary is 128 MiB. The pair therefore creates 256 MiB of transient allocation and roughly 512 MiB of producer-write plus reduction-read traffic per layer, before accounting for the reduction kernels themselves. Across the default 24-layer model, this is a potentially material training cost that must first be confirmed by profiling.

## Hard viability gate

Before changing either kernel, profile an untouched full RWKV7 attention/model backward on B300. Attribute time, allocation, and launch traffic to the two producer-plus-`aten::sum` sequences.

Proceed only if the pair is at least 5% of RWKV7 attention-backward GPU time, or if measurements show that their temporary allocation/traffic is a concrete throughput or memory limiter capable of supporting the model-level acceptance target. If this gate fails, record a no-go against this issue without implementing a microbenchmark-only optimization.

## Bounded implementation

For each auxiliary backward path:

1. preserve every existing per-token gradient (`dk`, `da`, `dr`, `dv`, `dg`, and related outputs);
2. accumulate the persistent-parameter contribution over a producer time tile in fp32;
3. write one compact partial per independent producer tile and parameter channel; and
4. launch a deterministic Triton reducer that combines partials in a fixed order and casts once to the parameter dtype.

Do not use atomics or reduce accumulation precision. Preserve int64 global address arithmetic, tail masks, both `k_a` shapes, fused-k short/long behavior, its explicit `cu_seqlens` contract, and gate correction's 65,536-token slicing. If folding the contribution into an existing multi-output producer causes register spills or occupancy loss, a separate compact contribution producer is allowed within these same two ops; broad reduction abstractions are not.

The primary design is architecture-neutral and must retain existing backend behavior. It must not import, duplicate, or modify the unmerged SM103 helper or HGRN work from fork PR #3. Do not touch `chunk_rwkv7`, DPLR/common recurrence kernels, RWKV7 equations, layer/model APIs, or unrelated autotuning.

## Expected files

- `fla/ops/rwkv7/fused_k_update.py`
- `fla/ops/rwkv7/gate_output_correction.py`
- `benchmarks/ops/registry.py`
- `tests/ops/test_rwkv7.py` only in the oracle/benchmark-fixture commit described below

No other source module is in scope without a blocker review.

## Oracle and benchmark fixture

After the viability gate passes, make one fixture-only commit before kernel edits:

- Register `fused_k_rwkv7` from `fla.ops.rwkv7.fused_k_update` with flattened `[B,T,H*D]` inputs and a flat `[H*D]` parameter.
- Register `gate_output_correction` from `fla.ops.rwkv7.gate_output_correction` with its existing `[B,T,H*D]`, `[B,T,H,D]`, and `[H,D]` contracts.
- Set `output_is_tuple=False` and `test_file='tests/ops/test_rwkv7.py'` for both entries, and add realistic per-op default shapes without changing global benchmark shapes.
- Add one deterministic direct `fused_k_rwkv7` oracle case for the explicit long (`T > 512`) `cu_seqlens` path using uneven, non-multiple sequence lengths, a nonuniform upstream gradient, and full output/`dk`/`da`/`dka` comparison against `k_update_ref`. Exercise the existing `cu_seqlens_cpu` option in this fixture as well.

Run that fixture against the untouched implementation and freeze its commit SHA. From then on, `tests/ops/test_rwkv7.py`, eager references, shapes, tolerances, dtypes, numeric flags, and the registry fixture are immutable. The kernel commits use the fixture SHA as the repository-native benchmark base; the underlying source baseline remains the upstream SHA above.

The existing frozen tests already cover fused-k at `T=13/4096/8000`, fp32/bf16, and both flat and `[1,1,D]` `k_a`, plus gate correction at `B4/T4096/H64/D64` and the `T=65537` slice boundary. They compare outputs and all gradients, including `dka` and `drk`, against eager fp32 references. No tolerance widening, skipped gradient, precision relaxation, or test/reference rewrite is permitted.

Promotion gates, with no `-k` filtering:

```bash
python -m benchmarks.ops.verify --op fused_k_rwkv7 --base <fixture-sha> --modes fwd fwdbwd
python -m benchmarks.ops.verify --op gate_output_correction --base <fixture-sha> --modes fwd fwdbwd
```

Also run the dependent suites selected by:

```bash
python scripts/find_dependent_tests.py \
  fla/ops/rwkv7/fused_k_update.py \
  fla/ops/rwkv7/gate_output_correction.py \
  benchmarks/ops/registry.py \
  tests/ops/test_rwkv7.py
```

Run the selected RWKV7 layer/model/cache/context-parallel tests plus required lint, header, compile, and changed-file pre-commit checks.

## B300 benchmark plan

Use the registered auxiliary ops in `fwd` and `fwdbwd` with bf16 public/default and boundary shapes, including:

- `B4/T4096/H32/D64`
- `B8/T512/H32/D64`
- `B1/T32768/H32/D64`
- `B1/T65537/H8/D64`
- the existing `H64/D64` stress regime

Run both ops in the same-session comparison where practical:

```bash
python -m benchmarks.ops.run \
  --op fused_k_rwkv7 gate_output_correction \
  --base <fixture-sha> --modes fwd fwdbwd \
  --custom-shapes '<fixed JSON matrix>' --json <result-path>
```

Use independent pinned source roots/caches if the built-in editable-worktree comparison is contaminated. Keep inputs, dtype, warmup, repetitions, clocks, and statistical method identical. Report median, mean, standard deviation, minimum, p10, p90, and equal-weight geomean; never select best runs. Forward is unchanged and must remain statistically neutral.

For public impact, run the native RWKV7 training benchmark on identical B300 allocations for dense and `--varlen` workloads, starting with the default hidden/head configuration at `B4/T4096/context_len=4096`. Use fixed inputs/seeds, identical model depth, warmup, steps, and optimizer settings, and report median tokens/s plus peak allocated/reserved memory. Include a long-context memory-feasible control and an unmodified default-config run. If the exact 24-layer public workload is not memory-feasible on the baseline, report that fact and use the largest identical depth that fits rather than silently changing the comparison.

## NCU evidence

Collect full and source reports on B300 for each baseline producer, its PyTorch reduction, each compact producer, and each final reducer. Cover the public dense shape, a long/slice-boundary shape, and the flattened packed model path. Preserve the complete launch trace.

Compare total sequence time, launch count, DRAM/L2 bytes and throughput, L2 hit rate, compute/memory SOL, achieved occupancy, registers, local-memory spills, eligible warps, and dominant stalls. Confirm that the full token-sized fp32 temporary and its Torch reduction are absent, no atomics were introduced, all expected work launched, and the measured traffic reduction explains the timing result.

Store raw reports and logs under an ignored `profile/` directory with exact baseline/fixture/candidate SHAs, commands, Slurm job IDs, GPU capability, clocks/idle evidence, driver, CUDA, PyTorch, Triton, and NCU versions. Do not commit reports or run standalone checksum passes.

## Acceptance and stop rules

- Both full frozen gates and all relevant dependent tests are green under NaN poisoning, with exact output shapes/dtypes and every gradient checked.
- At `B4/T4096/H32/D64`, each 128 MiB full fp32 temporary is replaced by at most 16 MiB of compact partials, unless a stricter measured design is used; NCU must show at least 85% less associated producer-plus-consumer DRAM traffic.
- No atomics, nondeterministic accumulation, spill regression, or precision/tolerance/reference change.
- Each auxiliary op improves `fwdbwd` by at least 1.05x at the public shape; their combined long/public equal-weight geomean is at least 1.10x; no target shape regresses more than 3% outside measured noise.
- Forward remains within 1% noise.
- Median native RWKV7 training throughput improves at least 3% on the representative public dense workload and is non-negative within 1% on packed/long controls; report peak-memory deltas.
- NCU attributes the result to removed full-sequence reduction traffic rather than skipped work, clock drift, stale autotune state, or a shifted bottleneck.

If correctness requires a looser oracle, the hard viability gate fails, model gain is below 3%, a target shape regresses beyond the bound, or profiling cannot explain the win, record a no-go rather than broadening into recurrence kernels, atomics, more architectures, or benchmark-only specialization.

## Duplicate audit and risks

As of the baseline audit, upstream has no open RWKV7 performance issue or PR for these two reductions. Open RWKV7 PR #962 concerns generation prompt-prefill correctness; open DPLR PR #1069 concerns the recurrence backend. Historical PRs #379, #525, and #534 introduced these auxiliary kernels but left the full fp32 gradient buffers in current `main`.

The main risk is numerical drift from reduction reassociation, especially long bf16 parameter gradients. The mandatory fixed-order fp32 hierarchy and frozen eager-gradient gates decide correctness. Producer-local accumulation can raise register pressure, and the 65,536-token slicing must contribute exactly once to the final `r_k` gradient. The direct fused-k varlen path currently launches through chunk indices, so the compact-partial indexing must not double-count duplicated or inactive programs. These are bounded implementation concerns inside the two auxiliary ops, not reasons to change the public API.
